src/transport_ofi.c

```diff
@@ -744,6 +771,88 @@ int allocate_recv_cntr_mr(void)
     return ret;
 }
 
+static inline
+int allocate_recv_cntr_mr(void)
+{
+    int ret = 0;
+    uint64_t flags = 0;
+
+    /* ------------------------------------ */
+    /* POST enable resources for to EP      */
+    /* ------------------------------------ */
+
+    /* since this is AFTER enable and RMA you must create memory regions for
+     * incoming reads/writes and outgoing non-blocking Puts, specifying entire
+     * VA range */
+
+#if ENABLE_TARGET_CNTR
+    {
+        struct fi_cntr_attr cntr_attr = {0};
+
+        /* Create counter for incoming writes */
+        cntr_attr.events   = FI_CNTR_EVENTS_COMP;
+        cntr_attr.wait_obj = FI_WAIT_UNSPEC;
+
+        ret = fi_cntr_open(shmem_transport_ofi_domainfd, &cntr_attr,
+                           &shmem_transport_ofi_target_cntrfd, NULL);
+        OFI_CHECK_RETURN_STR(ret, "target CNTR open failed");
+
+#ifdef ENABLE_MR_RMA_EVENT
+        if (shmem_transport_ofi_mr_rma_event)
+            flags |= FI_RMA_EVENT;
+#endif /* ENABLE_MR_RMA_EVENT */
+    }
+#endif
+
+#ifdef USE_FI_HMEM
+    if (shmem_external_heap_pre_initialized) {
+        ret = ofi_mr_reg_external_heap();
+        OFI_CHECK_RETURN_STR(ret, "OFI MR registration with HMEM failed");
+    }
+#endif
+    ret = ofi_mr_reg_bind(flags);
+    OFI_CHECK_RETURN_STR(ret, "OFI MR registration failed");
+
+    return ret;
+}
+
+#ifdef USE_FI_HMEM
+static
+int publish_external_mr_info(void)
```

Comment:
For my understanding, are **publish_external_mr_info** and **publish_mr_info** required to enable PEs to know the heap base (either external or internal) of other PEs?

Reply:
Yes.. we publish these through the PMI key value store, IIRC.